glommio/src/channels/channel_mesh.rs

```diff
@@ -0,0 +1,312 @@
+            executor_id: Local::id(),
+            sender,
+        }
+    }
+}
+
+type ChannelProvider<T> = (Cell<SenderProvider<T>>, Cell<ReceiverProvider<T>>);
+
+type ChannelProviders<T> = Vec<Vec<ChannelProvider<T>>>;
+
+#[derive(Clone)]
+pub struct MeshBuilder<T: Send + Copy> {
+    nr_peers: usize,
+    notifiers: Arc<RwLock<Vec<Notifier>>>,
+    providers: Arc<Vec<Vec<ChannelProvider<T>>>>,
+}
+
+unsafe impl<T: Send + Copy> Send for MeshBuilder<T> {}
+
+impl<T: Send + Copy> Debug for MeshBuilder<T> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        f.write_fmt(format_args!(
+            "MeshBuilder {{ nr_peers: {} }}",
+            self.nr_peers
+        ))
+    }
+}
+
+impl<T: 'static + Send + Copy> MeshBuilder<T> {
+    pub fn new(nr_peers: usize, channel_size: usize) -> Self {
+        MeshBuilder {
+            nr_peers,
+            notifiers: Arc::new(RwLock::new(Vec::new())),
+            providers: Arc::new(Self::build_providers(nr_peers, channel_size)),
+        }
+    }
+
+    fn build_providers(nr_peers: usize, channel_size: usize) -> ChannelProviders<T> {
+        (0..nr_peers)
+            .map(|i| {
+                (0..nr_peers)
+                    .map(|j| Self::build_provider(i, j, channel_size))
+                    .collect()
+            })
+            .collect()
+    }
+
+    fn build_provider(from: usize, to: usize, channel_size: usize) -> ChannelProvider<T> {
+        if from == to {
+            let (sender, receiver) = local_channel::new_bounded(channel_size.max(1));
+            let sender_provider = SenderProvider::Local(sender);
+            let receiver_provider = ReceiverProvider::Local(receiver);
+            (Cell::new(sender_provider), Cell::new(receiver_provider))
+        } else {
+            let (sender, receiver) = shared_channel::new_bounded(channel_size);
+            let sender_provider = SenderProvider::Shared(sender);
+            let receiver_provider = ReceiverProvider::Shared(receiver);
+            (Cell::new(sender_provider), Cell::new(receiver_provider))
+        }
+    }
+
+    async fn join(self) -> (Senders<T>, Receivers<T>) {
+        let mut notifiers = self.notifiers.write().unwrap();
+
+        if notifiers.len() == self.nr_peers {
+            panic!("The channel mesh is full.");
+        }
+
+        let index = notifiers
+            .binary_search_by(|n| n.executor_id.cmp(&Local::id()))
+            .expect_err("Should not join a mesh more than once.");
+
+        let peer_id = if notifiers.len() == self.nr_peers - 1 {
+            notifiers.insert(index, Notifier::new(None));
+            for (peer_id, notifier) in notifiers.iter_mut().enumerate() {
+                if let Some(sender) = notifier.sender.take() {
+                    sender.connect().await.send(peer_id).await.unwrap();
+                }
+            }
+            drop(notifiers);
+            index
+        } else {
+            let (sender, receiver) = shared_channel::new_bounded(1);
```

Comment:
if I understand correctly, this effectively means that only one message at a time can be sent, right ?

Full disclosure, I have a similar implementation in my project using glommio internally. It's fairly specific to what we are doing so I didn't consider moving to glommio. Yours is more generic so that's good!! (I don't need a full mesh)

However it is hard to achieve  good throughput with a single slot in the sender.

Reply:
Ok, this seems to exist only for notifications. I still haven't understood the role of notifications, though.

Reply:
The notification exists because peers need to wait for the last one to join to get their final order (peer id) in the mesh. Another way to implement this could be a polling loop to check the fullness of the mesh with a short sleep between each iteration. It could be simpler than the notification way.

Reply:
I don't mind the notification system. Just wanted to understand its role.